README.md
EpiKnet Idle RPG
=====

La documentation officielle se trouve sur [notre wiki](http://www.eirpg.com).

Vous pouvez signaler un bogue sur le [traqueur de problèmes de Github](https://github.com/ohmer1/eirpg/issues).

Retrouvez-nous sur #IRPG-CHAT sur le réseau EpiKnet pour discuter avec nous. :)

Bon idle !


doc/Changelog
Changelog
=========================================

Légende
-----------------
+ ajout
- suppression
* modification
! correction de bug


Version 0.6.1
15 septembre 2015
-----------------
+ Ajout de la commande AIDE/HELP (Luvwahraan)


Version 0.6.0
22 novembre 2012
-----------------
! Correction d'un message des mains de Dieu dans le fichier d'exemple SQL. [doc/textes_calamites_maindedieu.sql]
! Correction d'une autre anomalie empêchant la reconnexion automatique au serveur IRC lorsque le serveur n'est pas
    joignable. [clsIRC]
! Correction de la fonction start() qui ne tentait pas la reconnexion au serveur IRC en boucle, ce qui provoquait
    des erreurs en appellant la méthode boucle() de la classe IRC si la connexion IRC n'était pas rétablie
    immédiatement. [irpg.php]
! Correction d'une erreur dans la méthode connexion() provoquant la création d'un nouveau socket sans fermer au
    préalable le précédent s'il existe. [clsIRC]
! Correction de la méthode connexion() qui ne fermait pas le socket quand la connexion échouait ce qui ne permettait
    pas de savoir si la connexion était active. [clsIRC]
! Correction d'un bug dans la méthode boucle() qui ne s'interrompait pas si le socket n'existait pas. [clsIRC]


Version 0.6.0-rc6
13 novembre 2012
-----------------
! Correction de l'ordre des objets, ceux de base devant être insérés avant les objets uniques afin de satisfaire
    les relations SQL. [doc/liste_objets.sql]
! Correction de la méthode alog() de la classe IRPG pour prendre en compte le passage des fichiers en UTF-8. [clsIRPG]
+ Ajout d'un fichier SQL contenant les textes des calamités et des mains de Dieu
    manquants. [doc/textes_calamites_maindedieu.sql]
+ Ajout de la table Modules au fichier d'initialisation de la base de données. [doc/irpg.sql]
+ Prise en compte du paramètre charset dans la classe IRC pour convertir les commandes envoyées au serveur
    avec l'encodage désiré. [clsIRC]
* Conversion de tous les fichiers en UTF-8.
* Modification de l'encodage par défaut. [irpg.conf.dist]
* Modification du fichier SQL afin de spécifier UTF-8 comme encodage. [doc/irpg.sql]
* Modification des identifiants des objets pour correspondre à l'ordre d'insertion. [doc/liste_objets.sql]
* Modification du fichier SQL pour insérer tous les objets en une requête. [doc/liste_objets.sql]
* Modification de la méthode boucle() de la classe IRC pour tenter de convertir en UTF-8 les messages entrants qui
    seraient en ISO-8859-15. [clsIRC]


Version 0.6.0-rc5
10 novembre 2012
-----------------
! Correction d'un bug dans la liste affichée des participants à une quête lorsqu'il n'y en a plus qu'un. [mod_quests]
! Correction d'un bug faisant croire qu'une quête du survivant a été lancée même s'il n'y avait pas assez de joueurs
    pour y participer. [mod_quests]


Version 0.6.0-rc4
30 octobre 2012
-----------------
! Correction d'un bug qui provoquait le passage au niveau suivant d'un personnage fantôme lorsque la requête
    récupérant les personnages échoue  typiquement lorsque la base de donnée n'a pas été initialisée. [mod_idle]
! Correction d'une erreur dans le fichier de configuration d'exemple à cause d'un guillemet manquant. [irpg.conf.dist]
! Correction de quelques fautes dans le message de bienvenue de base. [irpg.conf.dist]


Version 0.6.0-rc3
25 octobre 2012
-----------------
! Correction d'erreurs de type E_STRICT provoquées par l'utilisation de la fonction mktime() sans
    arguments. [clsIRC, mod_core]
! Correction d'un bug empêchant la reconnexion automatique au serveur IRC en cas de perte de la connexion. [clsIRC]


Version 0.6.0-rc2
24 octobre 2012
-----------------
* Modification de la commande COMBAT afin de ne plus demander à l'utilisateur le nom de son personnage lorsqu'il
    n'en a qu'un. [mod_batailles]
* Correction de quelques fautes restantes. [doc/textes_mod_quests.sql, mod_objets, mod_top10]


Version 0.6.0-rc1
18 octobre 2012
-----------------
! Correction de convSecondes() qui utilisait le pluriel avec les durées inférieures à 1 jour. [clsIRPG]
! Correction de diverses erreurs : E_NOTICE, E_DEPRECATED, E_WARNING. [clsDb, clsIRC, clsIRPG]
! Correction de la commande admin SAY qui affichait des antislashes d'échappement. [mod_admin]
! Correction de la fonction alog() pour prendre en compte l'encodage spécifié et à défaut, utiliser l'encodage
    ISO-8859-15. [clsIRPG]
! Correction et amélioration de la gestion des quêtes parallèles qui écrasaient les participants de celle déjà en
    cours. [mod_quests]
! Correction de la commande admin QUESTSTART qui pouvait échouer sans afficher d'erreur lorsque le nombre
    d'utilisateurs requis n'était pas atteint. [mod_quests]
! Correction d'un bug ne permettant qu'à un seul des personnages par joueur de pouvoir participer à une
    quête. [mod_quests]
! Correction de la connexion SQL qui n'était pas fermée en cas d'interruption de la liaison avec le serveur,
    empêchant la reconnexion automatique. [clsDB]
! Correction des modes d'IRPG qui n'étaient pas activés sur un serveur IRC sous InspIRCd. [clsIRC]
! Correction d'une faille lors du traitement des données reçues pouvant mener à deux commandes corrompues si la
    taille maximale du buffer a été atteinte, au lieu de rassembler la commande lors de la réception du prochain lot
    de données. [clsIRC]
! Correction d'un bug dans onEndWho() qui empêchait l'ajout des utilisateurs dans la base de données si la fonction
    avait déjà été appelée auparavant. [clsIRC]
! Correction du type du champ Next dans le fichier de la base SQL, qui empêchait de passer au niveau suivant si le
    temps restant devait devenir négatif, en unsigned. [doc/irpg.sql]
! Correction d'une erreur lors de l'appel à la fonction cmdRegister2 : l'argument parrain ne lui était pas

lib/clsIRPG.php
<?php

/*
 * EpiKnet Idle RPG (EIRPG)
 * Copyright (C) 2005-2012 Francis D (Homer), cedricpc & EpiKnet
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. if not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Classe IRPG; classe très large qui regroupe plusieurs
 * fonctions utilisées par le bot
 *
 * @author    Homer
 * @author    cedricpc
 * @created   Lundi    30 Mai       2005
 * @modified  Mardi    13 Novembre  2012 @ 02:25 (CET)
 */
class IRPG
{
    ///////////////////////////////////////////////////////////////
    // Variables privées
    ///////////////////////////////////////////////////////////////
    var $config;   //Paramètres de configuration chargés en mémoire
    var $mod;      //Modules chargés
    var $modules;  //Noms des modules chargés
    var $pause;    //Indique si le jeu est en pause ou non
    var $ignoresN; //Liste des nicks ignorés en mémoire
    var $ignoresH; //Liste des hosts ignorés en mémoire
    ///////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////


    ///////////////////////////////////////////////////////////////
    // Méthodes privées, même si PHP s'en fou !
    ///////////////////////////////////////////////////////////////

    /**
     * Valide une section du fichier de configuration
     *
     * @author Homer
     * @author cedricpc
     * @created 1er juin 2005
     * @modified 19 Avril 2010
     * @return boolean - true si la config est OK, false autrement
     */
    function validationConfig($section, $keys, $keys_opt = null)
    {
        $config = parse_ini_file('irpg.conf', true);

        $this->config[$section] = array();
        $keys     = (array) $keys;
        $keys_opt = (array) $keys_opt;

        //On traite les clés obligatoires
        foreach ( $keys as $key) {
            if (empty($key)) {
                continue;
            }

            $value = (isset($config[$section][$key]) ? $config[$section][$key] : '');
            if ($value != '') {
                $this->config[$section][$key] = $value;
            } else {
                return false;
            }
        }

        //Ensuite, les clés optionnelles
        foreach ($keys_opt as $key) {
            if (empty($key)) {
                continue;
            }

            $this->config[$section][$key] = (isset($config[$section][$key]) ? $config[$section][$key] : '');
        }

        //On est évidemment pas en pause !
        $this->pause = false;

        //Si on s'est rendu ici c'est que tout fonctionne !
        return true;
    }

    ///////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////

    ///////////////////////////////////////////////////////////////
    // Méthodes publiques
    ///////////////////////////////////////////////////////////////

    /**
     * Constructeur; valide et charge la configuration
     *
     * @author Homer
     * @created 30 mai 2005
     * @modified 1er juin 2005
     * @return boolean - true si la config est OK, false autrement
     */
    function init()
    {
        $this->alog("Lecture du fichier de configuration...", true);

        //Traitement de la section DB
        $keys     = array("host", "login", "password", "base"); //clés obligatoires
        $keys_opt = array("prefix", "charset");                 //clés optionnelles

        $ok = $this->validationConfig("SQL", $keys, $keys_opt);
        if (!$ok) {
            return false;
        }

        //Traitement de la section IRC
        $keys     = array(
            "server", "port", "channel", "nick", "altnick", "username", "realname", "modes"
        );                                                                 //clés obligatoires
        $keys_opt = array("password", "nspass", "bind", "key", "charset"); //clés optionnelles

        $ok = $this->validationConfig("IRC", $keys, $keys_opt);
        if (!$ok) {
            return false;
        }

        //Traitement de la section IRPG
        $keys     = array("admin", "debug", "background", "purge", "version", "modules"); //clés obligatoires
        $keys_opt = array("charset");                                                     //clés optionnelles

        $ok = $this->validationConfig("IRPG", $keys, $keys_opt);
        if (!$ok) {
            return false;
        }

        $this->ignoresH = array();
        $this->ignoresN = array();

        //Si on s'est rendu ici c'est que tout est OK :)
        return true;
    }

///////////////////////////////////////////////////////////////

    /**
     * Charge les modules
     *
     * @author Homer
     * @created 20 juin 2005
     * @modified 20 juin 2005
     * @param clsIRC      - Référence à l'objet IRC
     * @param clsIRPG     - Référence à l'objet IRPG
     * @return none
     */
    function loadModules()
    {
        global $irc, $irpg;

        //Chargement des modules
        $this->modules = $this->readConfig("IRPG", "modules");
        $this->modules = explode(',', $this->modules);

        //On vérifie que les modules existent
        $i = 0;
        while ($i != count($this->modules)) {
            if (!file_exists("modules/mod_" . $this->modules[$i] . ".php")) {
                die("Le module mod_" . $this->modules[$i] . " n'existe pas\n");
            } else {
                $this->alog("Chargement du module mod_" . $this->modules[$i] . "...", true);
                include "modules/mod_" . $this->modules[$i] . ".php";
                $this->mod[$this->modules[$i]] = new $this->modules[$i]();
                $this->mod[$this->modules[$i]]->loadModule();
            }
            $i++;
        }
    }

///////////////////////////////////////////////////////////////

    /**
     * Charge un module
     *
     * @author Homer
     * @created 22 juin 2005
     * @modified 22 juin 2005
     * @param nom         - Nom du module à charger (sans le prefixe mod_)
     * @return boolean    - true si module chargé, false autrement
     */
    function loadModule($nom)
    {
        if (!file_exists("modules/mod_" . $nom . ".php")) { //On vérifie que le module existe
            return false;
        } elseif (in_array($nom, $this->modules)) { //On s'assure que le module ne soit pas déjà chargé
            return false;
        } else {
            include_once "modules/mod_" . $nom . ".php"; //TODO: rechercher le module sur REHASH (??)
            $this->modules[] = $nom;
            $this->mod[$nom] = new $nom;
            $this->mod[$nom]->loadModule();
            return true;
        }
    }

///////////////////////////////////////////////////////////////

    function unloadModule($nom)
    {
        //On vérifie si le module existe
        if (!in_array($nom, $this->modules)) {
            return false; //module inexistant
        } else {
            $i = 0;
            foreach ($this->mod as $nomModule => $leModule) {
                $y = 0;
                while ($y != count($this->mod[$nom]->depend)) {
                    if ($this->mod[$nomModule]->depend[$y] == $nom . "/" . $this->mod[$nom]->version) {
                        return false; // On ne peut décharger ce module car il est requis
                        break;        // par un autre module actuellement chargé
                    }
                    $y++;
                }

                $i++;
            }

            //On peut maintenant décharger le module
            //On appel l'évènement unloadmodule() avant de décharger
            $this->mod[$nom]->unloadmodule();
            //On retire le module des tableaux de modules
            unset($this->mod[$nom]);
            unset($this->modules[(array_search($nom, $this->modules))]);
            $this->modules = array_values($this->modules); //reset des indices du tableau
            return true;
        }
    }

///////////////////////////////////////////////////////////////

    /**
     * Config; interprète le fichier de configuration
     *
     * @author Homer
     * @author cedricpc
     * @created 30 mai 2005
     * @modified 19 Avril 2010
     * @param section    - Section du fichier de config (SQL, IRC, IRPG...)
     * @param key        - Élément du fichier que l'on recherche (server, port, login..)'
     * @param fromFile   - Si vrai lis la configuration du fichier, sinon des données en mémoire
     * @return string    - La valeur contenu dans le fichier de configuration
     */
    function readConfig($section, $key, $fromFile = false)
    {
        if ($fromFile) {
            $config = parse_ini_file("irpg.conf", true);
            return (isset($config[$section][$key]) ? $config[$section][$key] : null);
        } else {
            //Retourne ce qui a été préablement chargé en mémoire
            return (isset($this->config[$section][$key]) ? $this->config[$section][$key] : null);
        }
    }

///////////////////////////////////////////////////////////////

    /**
     * Vérifie si les dépendances d'un
     * module sont satisfaites
     *
     * @author Homer
     * @created 20 juin 2005
     * @modified 22 juin 2005
     * @param dep[]    - Tableau contenant les modules nécessaires
     * @return boolean - True si dépendances satisfaites, false autrement.
     */
    function checkDepd($dep)
    {
        $i = 0;
        while ($i != count($dep)) {
            //Nom & version de la dépendance requise
            $module        = explode('/', $dep[$i]);
            $nomModule     = $module[0];
            $versionModule = $module[1];

            if (empty($nomModule)) {
                return true;
            }

            //On vérifie si le module est chargé
            if (!in_array($nomModule, $this->modules)) {
                return false;
            } else {
                //On vérifie que la version du module est suffisante

                //Version requise par le module
                $versionModule = explode('.', $versionModule);
                $vr_majeur     = $versionModule[0];
                $vr_mineur     = $versionModule[1];
                $vr_revision   = $versionModule[2];

                //Version actuelle du module
                $versionActuelle = explode('.', $this->mod[$nomModule]->version);
                $va_majeur       = $versionActuelle[0];
                $va_mineur       = $versionActuelle[1];
                $va_revision     = $versionActuelle[2];

                if ($va_majeur > $vr_majeur) {
                    //return true;
                } elseif ($va_majeur < $vr_majeur) {
                    return false;
                } elseif ($va_mineur > $vr_mineur) {
                    //return true;
                } elseif ($va_mineur < $vr_mineur) {
                    return false;
                } elseif ($va_revision >= $vr_revision) {
                    //return true;
                } elseif ($va_revision < $vr_revision) {
                    return false;
                } else {
                    return false;
                }
            }
            $i++;
        }

        return true;
    }

///////////////////////////////////////////////////////////////

    function alog($msg, $print = false)
    { //Gestion des logs et de l'affichage des info de débuguage
        $date = date("j-m-Y H:i:s");
        if ((!$this->readConfig("IRPG", "background") || ($print))) {
            $charset = $this->readConfig("IRPG", "charset");
            print iconv('UTF-8', ($charset ? $charset : 'ISO-8859-15') . '//TRANSLIT', "[$date] $msg\n");
        }
        $flog = fopen("irpg.log", "a+");
        fwrite($flog, "[$date] " . $msg . "\n");
        fclose($flog);
    }

///////////////////////////////////////////////////////////////

    function getUsernameByNick($nick, $uid = false)
    {
        $username = array_search($nick, $this->mod["core"]->users);

        if (($uid) && ($username)) {
            global $db;
            $table = $db->prefix . "Utilisateurs";
            $uid = $db->getRows("SELECT Id_Utilisateurs FROM $table WHERE Username = '$username'");

            return array($username, (empty($uid) ? null : $uid[0]["Id_Utilisateurs"]));
        } else {
            return $username;
        }
    }

///////////////////////////////////////////////////////////////

    function getNickByUID($uid)
    {
        global $db;

        $tbIRC   = $db->prefix . "IRC";
        $tbPerso = $db->prefix . "Personnages";

        $nick = $db->getRows("SELECT Nick FROM $tbIRC WHERE Pers_Id = (SELECT Id_Personnages FROM $tbPerso
                              WHERE Util_Id='$uid' LIMIT 0,1)");

        return (empty($nick) ? false : $nick[0]["Nick"]);
    }

 ///////////////////////////////////////////////////////////////

    function getUsernameByUID ($uid)
    {
        global $db;

        $tbUtil = $db->prefix . "Utilisateurs";
        $username = $db->getRows("SELECT Username FROM $tbUtil WHERE Id_Utilisateurs = '$uid' LIMIT 0,1");

        return (empty($username) ? false : $username[0]['Username']);
    }

///////////////////////////////////////////////////////////////

    function getNomPersoByPID ($pid)
    {
        global $db;

        $tbPerso = $db->prefix . "Personnages";
        $result = $db->getRows("SELECT Nom FROM $tbPerso WHERE Id_Personnages = '$pid'");

        return (empty($result) ? false : $result[0]['Nom']);
    }

///////////////////////////////////////////////////////////////

    function getPIDByPerso ($perso)
    {
        global $db;

        $tbPerso = $db->prefix . "Personnages";
        $result = $db->getRows("SELECT Id_Personnages FROM $tbPerso WHERE Nom = '$perso'");

        return (empty($result) ? false : $result[0]['Id_Personnages']);
    }

///////////////////////////////////////////////////////////////

    function getPersoByUsername($username)
    {
    }

///////////////////////////////////////////////////////////////

    function getUIDByPID ($pid)
    {
        global $db;

        $tb  = $db->prefix . "Personnages";
        $res = $db->getRows("SELECT Util_Id FROM $tb WHERE Id_Personnages='$pid'");

        return (empty($res) ? false : $res[0]['Util_Id']);
    }

///////////////////////////////////////////////////////////////

    function getUIDByUsername($username)
    {
        global $db;

        $tbUtil = $db->prefix . "Utilisateurs";
        $uid = $db->getRows("SELECT Id_Utilisateurs FROM $tbUtil WHERE Username='$username'");

        return (empty($uid) ? false : $uid[0]['Id_Utilisateurs']);
    }

///////////////////////////////////////////////////////////////

    function convSecondes($sec)
    {
        if ($sec == 0) {
            return "00:00:00";
        } else {
            return sprintf("%d jour%s, %02d:%02d:%02d", $sec / 86400, $sec < 172800 ? "" : "s",
                ($sec % 86400) / 3600, ($sec % 3600) / 60, $sec % 60
            );
        }
    }

///////////////////////////////////////////////////////////////

    function getAdminLvl($uid)
    {
        //Retourne le niveau d'accès admin d'un utilisateur
        global $db;

        $tbUtil = $db->prefix . "Utilisateurs";
        $resultat = $db->getRows("SELECT Admin FROM $tbUtil WHERE Id_Utilisateurs = '$uid'");

        return (empty($resultat) ? 0 : $resultat[0]['Admin']);
    }

///////////////////////////////////////////////////////////////

    function Log($pid, $type, $modif = 0, $d1 = "", $d2 = "", $d3 = "")
    {
        //Ajout dans la table Logs
        global $db;

        $tbLogs = $db->prefix . "Logs";
        if ($pid == null) {
            $db->req("INSERT INTO $tbLogs (`Pers_Id`, `Date`, `Type`, `Modificateur`, `Desc1`, `Desc2`, `Desc3`)
                      VALUES (NULL, NOW(), '$type', '$modif', '$d1', '$d2', '$d3')");
        } else {
            $db->req("INSERT INTO $tbLogs (`Pers_Id`, `Date`, `Type`, `Modificateur`, `Desc1`, `Desc2`, `Desc3`)
                      VALUES ('$pid', NOW(), '$type', '$modif', '$d1', '$d2', '$d3')");
        }
    }

///////////////////////////////////////////////////////////////

    function userExist($user)
    {
        global $db;

        $table = $db->prefix . "Utilisateurs";
        $r = $db->req("SELECT Username FROM $table WHERE Username='$user'");
        if (mysql_num_rows($r) != 0) {
            return true;
        }
        return false;
    }

///////////////////////////////////////////////////////////////

    function persoExist($perso)
    {
        global $db;

        $table = $db->prefix . "Personnages";
        $r = $db->req("SELECT Nom FROM $table WHERE Nom='$perso'");
        if (mysql_num_rows($r) != 0) {
            return true;
        }
        return false;
    }

    function equipeExist($equipe)
    {
        global $db;

        $table = $db->prefix . "Equipes";
        $r = $db->req("SELECT Name FROM $table WHERE Name='$equipe'");
        if (mysql_num_rows($r) != 0) {
            return true;
        }
        return false;
    }

///////////////////////////////////////////////////////////////

    function lireIgnores()
    {
        // lecture de la liste d'ignore en mémoire
        $this->alog("Lecture de la liste des ignores...");
        $f = fopen("ignores.list", "r");
        while (!feof($f)) {
            $ligne = fgets($f);

            if (substr($ligne, 0, 4) == "NICK") {
                $nick = explode(":", $ligne);
                $nick = trim($nick[1]);
                $this->ignoresN[] = $nick;
                $this->alog("Le pseudo $nick est ignoré...");
            } elseif (substr($ligne, 0, 4) == "HOST") {
                $host = explode(":", $ligne);
                $host = trim($host[1]);
                $this->ignoresH[] = $host;
                $this->alog("L'host $host est ignoré...");
            }
        }
        fclose($f);
    }

    function getIgnoresN()
    {
        return $this->ignoresN;
    }

    function getIgnoresH()
    {
        return $this->ignoresH;
    }

    ///////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////
}
?>
